File: examples/grpc_examples/typescript/simple_execution.ts
/**
 * Simple ExecutePipeline example
 * 
 * Demonstrates:
 * - Connecting to the gRPC service
 * - Checking version compatibility
 * - Executing a simple calculator pipeline
 * - Handling results and errors
 */

import { RemoteMediaClient } from '../../../nodejs-client/src/client';
import { PipelineManifest, ManifestMetadata, NodeManifest, ExecuteRequest } from '../../../nodejs-client/generated-types/execution_pb';

async function main() {
  // Create client (no auth for local development)
  const client = new RemoteMediaClient('localhost:50051');
  
  try {
    // Connect
    console.log('Connecting to gRPC service...');
    await client.connect();
    
    // Check version
    const version = await client.getVersion();
    console.log(`✅ Connected to service v${version.getVersionInfo()?.getProtocolVersion()}`);
    console.log(`   Runtime version: ${version.getVersionInfo()?.getRuntimeVersion()}`);
    
    const nodeTypes = version.getVersionInfo()?.getSupportedNodeTypesList() || [];
    console.log(`   Supported nodes: ${nodeTypes.slice(0, 5).join(', ')}...`);
    
    // Create simple calculator pipeline
    console.log('\n=== Executing Calculator Pipeline ===');
    
    const metadata = new ManifestMetadata();
    metadata.setName('simple_calculator');
    metadata.setDescription('Add 5 to input value');
    metadata.setCreatedAt('2025-10-28T00:00:00Z');
    
    const calcNode = new NodeManifest();
    calcNode.setId('calc');
    calcNode.setNodeType('CalculatorNode');
    calcNode.setParams('{"operation": "add", "value": 5.0}');
    calcNode.setIsStreaming(false);
    
    const manifest = new PipelineManifest();
    manifest.setVersion('v1');
    manifest.setMetadata(metadata);
    manifest.setNodesList([calcNode]);
    manifest.setConnectionsList([]);
    
    // Execute pipeline
    const request = new ExecuteRequest();
    request.setManifest(manifest);
    request.getDataInputsMap().set('calc', '{"value": 10.0}');
    request.setClientVersion('v1');
    
    const response = await client.executePipeline(request);
    
    if (response.hasResult()) {
      const result = response.getResult()!;
      console.log('✅ Execution successful');
      console.log('   Input: 10.0');
      console.log('   Operation: add 5.0');
      console.log(`   Result: ${result.getDataOutputsMap().get('calc')}`);
      console.log(`   Wall time: ${result.getMetrics()?.getWallTimeMs().toFixed(2)}ms`);
    } else {
      const error = response.getError()!;
      console.error(`❌ Error: ${error.getMessage()}`);
      process.exit(1);
    }
    
    // Try multiplication
    console.log('\n=== Executing Multiply Pipeline ===');
    
    calcNode.setParams('{"operation": "multiply", "value": 3.0}');
    metadata.setDescription('Multiply by 3');
    
    request.getDataInputsMap().set('calc', '{"value": 7.0}');
    
    const response2 = await client.executePipeline(request);
    
    if (response2.hasResult()) {
      const result = response2.getResult()!;
      console.log('✅ Execution successful');
      console.log('   Input: 7.0');
      console.log('   Operation: multiply 3.0');
      console.log(`   Result: ${result.getDataOutputsMap().get('calc')}`);
      console.log(`   Wall time: ${result.getMetrics()?.getWallTimeMs().toFixed(2)}ms`);
    }
    
    console.log('\n✅ All tests passed!');
    
  } catch (error) {
    console.error(`\n❌ Error: ${error}`);
    process.exit(1);
  } finally {
    await client.disconnect();
  }
}

main();

File: examples/grpc_examples/python/simple_execution.py
#!/usr/bin/env python3
"""
Simple ExecutePipeline example.

Demonstrates:
- Connecting to the gRPC service
- Checking version compatibility
- Executing a simple calculator pipeline
- Handling results and errors
"""

import asyncio
import sys
from pathlib import Path

# Add python-grpc-client to path
sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "python-grpc-client"))

from remotemedia_client import RemoteMediaClient, RemoteMediaError


async def main():
    # Create client (no auth for local development)
    client = RemoteMediaClient("localhost:50051")
    
    try:
        # Connect
        print("Connecting to gRPC service...")
        await client.connect()
        
        # Check version
        version = await client.get_version()
        print(f"✅ Connected to service v{version.protocol_version}")
        print(f"   Runtime version: {version.runtime_version}")
        print(f"   Supported nodes: {', '.join(version.supported_node_types[:5])}...")
        
        # Create simple calculator pipeline
        print("\n=== Executing Calculator Pipeline ===")
        manifest = {
            "version": "v1",
            "metadata": {
                "name": "simple_calculator",
                "description": "Add 5 to input value",
                "created_at": "2025-10-28T00:00:00Z"
            },
            "nodes": [
                {
                    "id": "calc",
                    "node_type": "CalculatorNode",
                    "params": '{"operation": "add", "value": 5.0}',
                    "is_streaming": False
                }
            ],
            "connections": []
        }
        
        # Execute pipeline
        result = await client.execute_pipeline(
            manifest=manifest,
            audio_inputs={},
            data_inputs={"calc": '{"value": 10.0}'}
        )
        
        # Display result
        print(f"✅ Execution successful")
        print(f"   Input: 10.0")
        print(f"   Operation: add 5.0")
        print(f"   Result: {result.data_outputs['calc']}")
        print(f"   Wall time: {result.metrics.wall_time_ms:.2f}ms")
        
        # Try multiplication
        print("\n=== Executing Multiply Pipeline ===")
        manifest["nodes"][0]["params"] = '{"operation": "multiply", "value": 3.0}'
        manifest["metadata"]["description"] = "Multiply by 3"
        
        result = await client.execute_pipeline(
            manifest=manifest,
            audio_inputs={},
            data_inputs={"calc": '{"value": 7.0}'}
        )
        
        print(f"✅ Execution successful")
        print(f"   Input: 7.0")
        print(f"   Operation: multiply 3.0")
        print(f"   Result: {result.data_outputs['calc']}")
        print(f"   Wall time: {result.metrics.wall_time_ms:.2f}ms")
        
        print("\n✅ All tests passed!")
        
    except RemoteMediaError as e:
        print(f"\n❌ Error: {e.message}")
        if e.error_type:
            print(f"   Error type: {e.error_type.name}")
        if e.failing_node_id:
            print(f"   Failing node: {e.failing_node_id}")
        sys.exit(1)
        
    finally:
        await client.disconnect()


if __name__ == "__main__":
    asyncio.run(main())
